### PR TITLE
eventDispatcher

### PR DESCRIPTION
**Pull request by nikoladjuric88**

### CODE REVIEW

File: src/eventDispatcher.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+function EventDispatcher () {
+	this.eventDispatcherMap = {};
+};
+
+EventDispatcher.prototype.bind = function(nameOfFunction, whichFunction) {
```
Comment (zikicm) [inline src/eventDispatcher.js]: Event dispatcher dispatches events, not functions:
`nameOfFunction` -> `eventName`
`whichFunction` -> `function` or `handler`

File: src/eventDispatcher.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	if(this.eventDispatcherMap[nameOfFunction]) {
+		this.eventDispatcherMap[nameOfFunction].push(whichFunction);
+	} else {
+		this.eventDispatcherMap[nameOfFunction] = [whichFunction];
+	}
+};
+
+EventDispatcher.prototype.trigger = function(nameOfFunction) {
+	for (var i = 0; i < this.eventDispatcherMap[nameOfFunction].length; i++) {
```
Comment (zikicm) [inline src/eventDispatcher.js]: What will happen in the following scenario:
```
var eventDispatcher = new EventDispatcher();
eventDispatcher.trigger("exampleEvent");
```
?

**Review by zikicm — COMMENTED** [review]
We are making progress. We'll use this pull request to lear everything about event dispatcher and its purpose. After you resolve current comments, we'll start adding new stuff.

File: src/eventDispatcher.js (hunk 2 above)
Comment (nikoladjuric88) [reply src/eventDispatcher.js]: In that case you have a error message in console : 'Uncaught TypeError: Cannot read property 'length' of undefined' because you didn't make function bind with this eventName.

File: src/calculator.js [hunk 3]
```diff
@@ -3,6 +3,14 @@
 var Precedence = require('./precedence');
 var Screen = require('./screen');
 var Memory = require('./memory');
+var EventDispatcher = require('./eventDispatcher');
```
Comment (zikicm) [inline src/calculator.js]: Why have you added this file to the pull request? This is just some debug / testing code which must not be a part of repositorium, revert the change.

File: src/eventDispatcher.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+function EventDispatcher() {
+    this.eventDispatcherMap = {};
+};
+
+EventDispatcher.prototype.bind = function(eventName, handler) {
```
Comment (zikicm) [inline src/eventDispatcher.js]: Similar to bind, we'll need several other functions
1) `unbind(eventHame, handler)`
2) `unbindAll()` - removes all listeners for every event name

Please add them.

File: src/eventDispatcher.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        for (var i = 0; i < this.eventDispatcherMap[eventName].length; i++) {
+            this.eventDispatcherMap[eventName][i]();
+        };
+    };
+};
+
+EventDispatcher.prototype.unbind = function(eventName, handler) {
+     if (this.eventDispatcherMap[eventName]) {
+           for (var i = 0; i < this.eventDispatcherMap[eventName].length; i++) {
+                if(this.eventDispatcherMap[eventName][i].toString() === handler.toString()) {
```
Comment (zikicm) [inline src/eventDispatcher.js]: This check does not make sense, why are you converting functions to a string? If you are using anonymous functions that wont even work. For example:
```
var f1 = function() { console.log("Hello from A") }
var f2 = function() { console.log("Hello from B") }
ed.bind("hello", f1);
ed.bind("hello", f2)
ed.unbind("hello", f1);
ed.trigger("hello");
// nothing will be printed here
```

Comment (zikicm) [reply src/eventDispatcher.js]: Do it simpler.

File: src/eventDispatcher.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
+function EventDispatcher() {
+    this.eventDispatcherMap = {};
+};
+
+EventDispatcher.prototype.bind = function(eventName, handler) {
```
Comment (zikicm) [inline src/eventDispatcher.js]: What would happen in the following scenario:
```
ed.bind("someEvent", null);
ed.trigger("someEvent");
```
?
You need to handle this case.

File: src/eventDispatcher.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+        this.eventDispatcherMap[eventName].push(handler);
+    } else {
+        this.eventDispatcherMap[eventName] = [handler];
+    }
+};
+
+EventDispatcher.prototype.trigger = function(eventName) {
+    if (this.eventDispatcherMap[eventName]) {
+        for (var i = 0; i < this.eventDispatcherMap[eventName].length; i++) {
+            this.eventDispatcherMap[eventName][i]();
```
Comment (zikicm) [inline src/eventDispatcher.js]: What would happen in the following scenario:
```
ed.bind("someEvent", function() { throw new Errror("Hello from error"); }
ed.bind("someEvent", function() { console.log("Hello from second function"); }
```
?

You need to handle this case.

File: src/eventDispatcher.js [hunk 8]
```diff
@@ -0,0 +1,35 @@
+function EventDispatcher() {
```
Comment (zikicm) [inline src/eventDispatcher.js]: You forgot to add `use strict`.

File: src/eventDispatcher.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    if (this.eventDispatcherMap[eventName]) {
+        for (var i = 0; i < this.eventDispatcherMap[eventName].length; i++) {
+            this.eventDispatcherMap[eventName][i]();
+        };
+    };
+};
+
+EventDispatcher.prototype.unbind = function(eventName, handler) {
+    if (this.eventDispatcherMap[eventName]) {
+        for (i in this.eventDispatcherMap[eventName]) {
```
Comment (zikicm) [inline src/eventDispatcher.js]: Use normal `for`.

File: src/eventDispatcher.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    }
+}
+
+EventDispatcher.prototype.trigger = function(eventName) {
+    if (this.eventDispatcherMap[eventName]) {
+        for (var i = 0; i < this.eventDispatcherMap[eventName].length; i++) {
+            try {
+                this.eventDispatcherMap[eventName][i]();
+            } catch (e) {
+                this.eventDispatcherMap[eventName].splice(i, 1);
```
Comment (zikicm) [inline src/eventDispatcher.js]: Great work, I'm really proud :)

One important thing, we use artificial examples here. In reality, a function may or may not throw an exception, depends from various combinations. At one time it may throw an error, where in another it may not. Having that in mind, you should not exclude them from array here. You should just catch an error and ignore it.

**Review by zikicm — APPROVED** [review]
Great work, please resolve my comment before we you merge.

File: src/eventDispatcher.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+                } catch (e) {
+
+                }
+            }
+        }
+    }
+
+    unbind(eventName, handler) {
+        if (this.eventDispatcherMap[eventName]) {
+            for (var i = 0; i < this.eventDispatcherMap[eventName].length; i++) {
```
Comment (zikicm) [inline src/eventDispatcher.js:32]: Instead of `var`, for local variables you re now going to use `let`. Investigate the difference please.

File: src/eventDispatcher.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+            }
+        }
+    }
+
+    trigger(eventName) {
+        if (this.eventDispatcherMap[eventName]) {
+            for (var i = 0; i < this.eventDispatcherMap[eventName].length; i++) {
+                try {
+                    this.eventDispatcherMap[eventName][i]();
+                } catch (e) {
```
Comment (zikicm) [inline src/eventDispatcher.js:23]: Feel free to use shorter form:
```
} catch (e) { ... }
```

**Review by zikicm — APPROVED** [review]
Great effort, resolve my comments and merge. As a next step, in A SEPARATE CODE REVIEW, add JSDoc description for every method. In product ready code, each method needs to have a full description, so that once revisited it is easy to understand what it does.

You have a package for Sublime which will auto construct skeleton for JSDoc. Find it and install.
